File: R/Real_estate_application.R
source("Real_estate_helpers.R")
library(rgdal)
library(rgeos)
library(maptools)
library(broom)

path <- "../real_estate_dataset/"
data <- readRDS(paste0(path,"scrape_all_2019_to_2020_raw_20200714.Rds"))
targets <- read.csv(paste0(path,"suburbs_list_20200714.csv"))[,2:3]

trans_path <- paste0(path, "trans/")
data_clean <- clean_data(data, date = "20200718", name = "FULL", save = F, path = path)
data_trans <- transform_data(data_clean, date = "20200718", name = "FULL", save = F, path = trans_path)

nrow(data_trans) # n = 13,291

# VISUALISING WITH MAP DATA  --------------------------------------
data_clean <- readRDS(paste0(path, "data_clean_FULL_20200718.Rds"))
data_trans <- readRDS(paste0(path, "trans/data_trans_FULL_20200718.Rds"))

dists <- data_clean$CBD
data_clean$dist_group <- factor(ceiling(dists/10))
X11(); boxplot(land ~ dist_group, data = data_clean)
hist(data_clean$land)

sample_ind <- sample(1:nrow(data_clean), 1000)
X11(); ggplot(data_clean[sample_ind,]) +
  geom_point(aes(x = CBD, y = price, size = land, colour = room))
1203/953
1824/1130

hist(data_clean$CBD)

# looks like everyone rushed to buy houses at start of covid.
barplot(table(data_trans$date))
# Zipf-looking suburb barplot
barplot(sort(table(data_trans$suburb)))

top_5_suburbs <- rev(sort(table(data_trans$suburb)))[1:5]

# VIC Suburb/Locality Boundaries - PSMA Administrative Boundaries
# -- Source: https://data.gov.au/dataset/ds-dga-af33dd8c-0534-4e18-9245-fc64440f742e/details
path2 <- "../../../real_estate_dataset/"
spdf <- readOGR(dsn= paste0(path2,"/vic_locality_polygon_shp"))
mapdata <- tidy_map_data(spdf, data_trans, targets)

# Based on the below plot, we dropped hillside and bellfield earlier

X11(); plot_map_with_suburb_labels(mapdata)
# Note that CBD is bimodal
histograms(data_trans)

# We remove this suburb cocoroc because it does not plot well (and only has one house)
data_clean <- data_clean[data_clean$suburb != "COCOROC",]
# Now compare close to far
mapdata_close <- tidy_map_data(spdf, data_trans[data_clean$CBD <= 25,], targets)
mapdata_far <- tidy_map_data(spdf, data_trans[data_clean$CBD > 25,], targets)
pdf(file="suburb_map.pdf",width=6,height=5)
plot_map_with_near_and_far(mapdata_close, mapdata_far)
dev.off()

# Now compare 2019 to 2020
mapdata2019 <- tidy_map_data(spdf, data_trans[data_trans$year == "2019",], targets)
mapdata2020 <- tidy_map_data(spdf, data_trans[data_trans$year == "2020",], targets)
X11(); plot_map(mapdata2019)
X11(); plot_map(mapdata2020)

###################################################################
# SHAPLEY VALUES ARE CALCULATED IN Julia/Real_estate_application.jl
###################################################################
date_condition <- function(date) {
  (date >= "2019-02-01" & date <= "2019-04-01") |
  (date >= "2020-02-01" & date <= "2020-04-01")
}
date_table_df <- date_counts(data_trans$date)
#pdf(file="dates_barplot.pdf",width=5,height=4)
plot_date_table(date_table_df)
#dev.off()

data_st <- split_and_transform(data_clean, date_condition = date_condition, dist = 25)
sum(sapply(data_st, nrow))
sapply(data_st, nrow)

m1 <- lm(price ~ CBD+images+land+school+station+room, data = data_st$near_2019)
m2 <- lm(price ~ CBD+images+land+school+station+room, data = data_st$far_2019)
m3 <- lm(price ~ CBD+images+land+school+station+room, data = data_st$near_2020)
m4 <- lm(price ~ CBD+images+land+school+station+room, data = data_st$far_2020)
summary(m1)$r.sq
summary(m2)$r.sq
summary(m3)$r.sq
summary(m4)$r.sq

#plot(m1)

pfx <- "data_shapley_"
shp_fourway <- read.csv(paste0(path, pfx, "fourway_split_20200718.csv"))
shp_year <- read.csv(paste0(path, pfx, "year_split_20200718.csv"))

# The data for this was transformed in four groups: by year/dist

shp_fourway$year <- factor(shp_fourway$year)
shp_fourway$type <- factor(shp_fourway$type) # Type is distance near/far
View(shp_fourway)
pdf(file="realestate_shapley2.pdf",width=5,height=4)
ggplot(data = shp_fourway) +
  geom_point(aes(x = feature, y = Shapley, shape = year)) +
  geom_crossbar(
    fatten=1, alpha=0.3, width=0.3, linetype=0,
    aes(y=Shapley, ymin = CIL, ymax = CIU, x = feature,
        colour = year, fill = year)) +
  facet_grid(type ~ .)
dev.off()

# The data for this was transformed in four groups: by year/dist
pdf(file="realestate_shapley.pdf",width=5,height=4)
ggplot(data = shp_fourway) +
  geom_point(aes(x = feature, y = Shapley, shape = type)) +
  geom_crossbar(
    fatten=1, alpha=0.3, width=0.3, linetype=0,
    aes(y=Shapley, ymin = CIL, ymax = CIU, x = feature,
        colour = type, fill = type)) +
  facet_grid(year ~ .)
dev.off()

# The data for this was transformed in two groups: by year
shp_year$year <- factor(shp_year$year)
ggplot(data = shp_year) +
  geom_point(aes(x = feature, y = Shapley)) +
  geom_crossbar(
    fatten=1, alpha=0.3, width=0.3, linetype=0,
    aes(y=Shapley, ymin = CIL, ymax = CIU, x = feature,
        colour = year, fill = year))

File: Julia/Real_estate_application.jl
using CSV, Distributions, DataFrames

# move to directory of current file
cd( @__DIR__ )

include("Shapley.jl")
date = "20200927"
path = "../real_estate_dataset/"
data = CSV.read(path*"trans/data_trans_FULL_20200718.csv")
features = [:CBD, :images, :land, :school, :station, :room]

function format_dataframe(df, features)
    X = convert(Matrix, df[:, features])
    y = Array{Float64}(df[!, :price])
    Z = hcat(y,X)
    return Z
end

function all_together_shapleys(data, year, features;
    date = "TEST", type = "TEST", save = false,
    path = "../real_estate_dataset/", bootstrap_CIs = false,
    α = 0.05, Nb = 1000
  )
  Z = format_dataframe(data, features)
  n = size(Z)[1]
  s, v = calc_shapley(Z)
  CIℓ, CIu = calc_CIs(s, v, n)
  result = DataFrame(Shapley = s, variance = v, CIL = CIℓ, CIU = CIu)
  if (bootstrap_CIs) 
    bootshap, bCI = calc_bootstrap_CIs(Z, n, Nb; α = α)
    result = hcat(result, bCI)
  end
  if (length(type) > 0) result[!,:type] .= type end
  if (length(year) > 0) result[!,:year] .= year end
  result[!,:feature] = features
  if (save)
    CSV.write(path*"data_shapley_"*type*year*"_"*date*".csv", result)
  end
  return result
end

function calc_bootstrap_CIs(Z, n, Nb; α = 0.05)
  println("Calculating bootstrap CIs")
  bootshap = Matrix{Float64}(undef, size(Z,2)-1, Nb)
  for r in 1:Nb
    Zᵣ = Z[sample(1:n, n),:]
    bootshap[:,r] = just_shapley(Zᵣ)
    if ( r % 100 == 0 ) println("Computed $r of $Nb Shapley vectors") end
  end
  qt(col) = quantile(skipmissing(col), [α/2, 1-α/2])
  bCI = DataFrame(mapslices(qt, bootshap; dims = 2))
  rename!(bCI,["bCIL","bCIU"])
  return bootshap, bCI
end

### Split just 2019/2020

# Note that 2019 and 2020 have been independently
# transformed using Yeo-Johnson. The transformation was done in R.
function year_split_shapleys(path, trans_path, features;
  save = false, date = "TEST")
  years = ["2019","2020"]
  results = DataFrame()
  for y in years
    df = CSV.read(path*trans_path*"data_trans_"*y*"_20200718.csv")
    s = all_together_shapleys(df, y, features; date = date, type = "")
    results = vcat(results, s)
  end
  if (save)
    CSV.write(path*"data_shapley_year_split_"*date*".csv", results)
  end
  return results
end

#year_split_shapleys(path, "trans/", features, save = true, date = "TEST")

### Split 2019/2020 and near/far approach

# Note that the near/far and 2019/2020 have been independently
# transformed using Yeo-Johnson. The transformation was done in R.
function fourway_split_shapleys(path, trans_path, features;
    save = false, date = "TEST", bootstrap_CIs = false,
    α = 0.05, Nb = 1000
  )
  years = ["2019","2020"]
  dists = ["near","far"]
  results = DataFrame()
  for y in years
    for d in dists
      df = CSV.read(path*trans_path*"data_trans_"*d*"_"*y*"_20200718.csv")
      s = all_together_shapleys(df, y, features; date = date, type = d, 
        bootstrap_CIs = bootstrap_CIs, α = α, Nb = Nb)
      results = vcat(results, s)
    end
  end
  if (save)
    CSV.write(path*"data_shapley_fourway_split_"*date*".csv", results)
  end
  return results
end

@time fourway_split_shapleys(
  path, "trans/", features, save = true, date = date, bootstrap_CIs = true, Nb = 1000)



#m = Matrix{Float64}(undef, 5, 10)
#for i in 1:10
#  m[:,i] .= rand(5)
#end 
#qtt(col) = quantile(col[map(x -> !isnan(x), col)], [0.05/2, 1-0.05/2])